src/Adapters/EventSource.Net46/Implementation/ActivityPathDecoder.cs

```diff
@@ -0,0 +1,194 @@
+    /// </remarks>
+    internal static class ActivityPathDecoder
+    {
+        /// <summary>
+        /// The encoding for a list of numbers used to make Activity  Guids.   Basically
+        /// we operate on nibbles (which are nice because they show up as hex digits).  The
+        /// list is ended with a end nibble (0) and depending on the nibble value (Below)
+        /// the value is either encoded into nibble itself or it can spill over into the
+        /// bytes that follow.   
+        /// </summary>
+        private enum NumberListCodes : byte
+        {
+            End = 0x0,             // ends the list.   No valid value has this prefix.   
+            LastImmediateValue = 0xA,
+            PrefixCode = 0xB,
+            MultiByte1 = 0xC,   // 1 byte follows.  If this Nibble is in the high bits, it the high bits of the number are stored in the low nibble.   
+                                // commented out because the code does not explicitly reference the names (but they are logically defined).  
+                                // MultiByte2 = 0xD,   // 2 bytes follow (we don't bother with the nibble optimzation
+                                // MultiByte3 = 0xE,   // 3 bytes follow (we don't bother with the nibble optimzation
+                                // MultiByte4 = 0xF,   // 4 bytes follow (we don't bother with the nibble optimzation
+        }
+
+        /// <summary>
+        /// Returns true if 'guid' follow the EventSource style activity ID for the process with ID processID.  
+        /// You can pass a process ID of 0 to this routine and it will do the best it can, but the possibility
+        /// of error is significantly higher (but still under .1%).
+        /// </summary>
+        public static unsafe bool IsActivityPath(Guid guid, int processID)
+        {
+            uint* uintPtr = (uint*)&guid;
+
+            uint sum = uintPtr[0] + uintPtr[1] + uintPtr[2] + 0x599D99AD;
+            if (processID == 0)
+            {
+                // We guess that the process ID is < 16 bits and because it was xored
+                // with the lower bits, the upper 16 bits should be independent of the
+                // particular process, so we can at least confirm that the upper bits
+                // match. 
+                return (sum & 0xFFFF0000) == (uintPtr[3] & 0xFFFF0000);
+            }
+
+            if ((sum ^ (uint)processID) == uintPtr[3])  
+            {
+                // This is the new style 
+                return true;
+            }
+            return sum == uintPtr[3];         // THis is old style where we don't make the ID unique machine wide.  
+        }        
+
+        /// <summary>
+        /// Returns a string representation for the activity path.  If the GUID is not an activity path then it returns
+        /// the normal string representation for a GUID. 
+        /// </summary>
+        public static unsafe string GetActivityPathString(Guid guid)
```

Comment:
Good point, I added a few tests for this method.

Full disclosure: I copied that code from PerfView. Could not find any unit tests to steal.